### PR TITLE
TRPO_GAE

### PR DESCRIPTION
**Pull request by sef007**
Added initial working implementation of TRPO.

### CODE REVIEW

**Comment by TheGogy** [comment]
Closes #16 
Closes #15

**Comment by TheGogy** [comment]
Might be an idea to create a separate directory in the repo that we can keep optimal agent weights so that we don't have to re-train.

File: src/agents/trpo/policy_network.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import torch
+import torch.nn as nn
+from torch.distributions import Normal
+from typing import List
+
+class PolicyNetwork(nn.Module):
+    """
+    a stochastic mapping for states to actions.
+    """
+    def __init__(self, state_dim: int, action_dim: int, max_action: float, hidden_sizes: List[int]):
```
Comment (TheGogy) [inline src/agents/trpo/policy_network.py]:
```suggestion
    def __init__(self, state_dim: int, action_dim: int, max_action: float, hidden_sizes: List[int], upper_bound: float, lower_bound: float):
```

File: src/agents/trpo/policy_network.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        """
+        policy network to be used as a module. 
+        """
+        super().__init__()
+        self.max_action = max_action
+
```
Comment (TheGogy) [inline src/agents/trpo/policy_network.py]:
```suggestion
    self.upper_bound = upper_bound
    self.lower_bound = lower_bound
```

File: src/agents/trpo/agent.py [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+# 9. Select Action 
+# 10. Update Value Function
+################################################################################        
+class TRPOAgent:
+    def __init__(self, state_dim: int, action_dim: int, max_action: float, device: torch.device):
+        """
+        initalizes all the hyper parameters and calls both networks 
+        """
+        # Determines capacity of the network
+        self.policy = PolicyNetwork(state_dim, action_dim, max_action, [64, 64]).to(device)
```
Comment (TheGogy) [inline src/agents/trpo/agent.py]:
```suggestion
        UPPER_BOUND = 2
        LOWER_BOUND = -20
        self.policy = PolicyNetwork(state_dim, action_dim, max_action, [64, 64], UPPER_BOUND, LOWER_BOUND).to(device)
```

File: src/agents/trpo/agent.py [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+                'actions': [], 
+                'rewards':[], 
+                'values': [],
+                'probabilitys': []
+            } 
+
+            state, _ = env.reset()
+            done = False
+
+            while done == False:
```
Comment (TheGogy) [inline src/agents/trpo/agent.py:107]:
```suggestion
            while not done:
```

File: src/agents/trpo/agent.py [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+                distribution  = self.policy.get_distribution(state_tensor)
+                action = distribution.sample()
+                log_prob = distribution.log_prob(action).sum( -1)
+
+                # start gradient 
+                torch.set_grad_enabled(True)
+                probabilitys = log_prob.cpu().item()
+
+                # get the flags
+                next_state, reward, terminated, truncated, dummy = env.step(action.cpu().numpy()[0])
```
Comment (TheGogy) [inline src/agents/trpo/agent.py]:
```suggestion
                next_state, reward, terminated, truncated, _ = env.step(action.cpu().numpy()[0])
```

File: src/agents/trpo/agent.py [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+
+                # increment loop 
+                total_step_count = total_step_count + 1
+                state = next_state
+
+                # end the episode
+                if (terminated == True) or (truncated == True): 
+                    done = True
+                else:
+                    done = False
```
Comment (TheGogy) [inline src/agents/trpo/agent.py:142]:
```suggestion
                # end the episode
                done = terminated or truncated
```

File: src/agents/trpo/agent.py [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+                # GAE advantage: last_gae = delta + gamma * lambda * last_gae
+                advantages[t] = delta + self.GAMMA * self.GAE_LAMBDA * last_gae
+            else:
+                # last step so no TD error. 
+                # GAE advantage = reward + gamma * lambda * value
+                delta = rewards[t] + self.GAMMA * (values[t] * -1)
+                advantages[t] = delta + self.GAMMA * self.GAE_LAMBDA * last_gae
+            
+            last_gae = advantages[t]
+            t = t -1  # iterate backwards
```
Comment (TheGogy) [inline src/agents/trpo/agent.py]:
```suggestion
            t -= 1 # Iterate backwards
```

File: src/agents/trpo/agent.py [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+        grads = torch.autograd.grad(
+            self.compute_kl_constraint(states, p_old), self.policy.parameters(),
+            create_graph=True,
+            retain_graph=True)
+        
+        # get in one d vector 
+        flat_grads_list = []
+        for grad in grads:
+            flat_grad = grad.view(-1)
+            flat_grads_list.append(flat_grad)
```
Comment (TheGogy) [inline src/agents/trpo/agent.py:255]:
```suggestion
        # get in one d vector 
        flat_grads_list = [g.view(-1) for g in grads]
```

File: src/agents/trpo/agent.py [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+        # approx fisher matrix = g * v 
+        g_v = (flat_grad_kl * vector)
+        grad_vector = (g_v).sum()
+
+        grad_grads = torch.autograd.grad(grad_vector, self.policy.parameters(), retain_graph=True)
+
+        # get in one d vector 
+        fisher_product_list = []
+        for grad in grad_grads:
+            fisher_product_list.append(grad.contiguous().view(-1))
```
Comment (TheGogy) [inline src/agents/trpo/agent.py:268]:
```suggestion
        # get in one d vector 
        fisher_product_list = [g.contiguous().view(-1) for g in grad_grads]
```

File: src/agents/trpo/agent.py [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+
+        # compute the surrogate loss using Step 3. and use it to the surrogate loss gradient 
+        grads = torch.autograd.grad(self.compute_surrogate_loss(states, actions, advantages, old_probabilitys),
+                                    self.policy.parameters())
+        
+        # convert back into a 1d vector again 
+        g = []
+        for grad in grads:
+            flat_grad = grad.view(-1)
+            g.append(flat_grad)
```
Comment (TheGogy) [inline src/agents/trpo/agent.py]:
```suggestion
        g = [grad.view(-1) for grad in grads]
```

File: src/agents/trpo/policy_network.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        """
+        works out the mean and std of the policy network 
+        """
+        # range for clamping, (has significant effect on convergence)
+        upper_bound = 2
+        lower_bound = -20
+        # forward_mean = max_action * tanh(1/N sum of state policies)
+        for_mean = self.max_action * torch.tanh(self.mean(self.policy(state)))  
+
+        clamped_log_std = self.log_std.clamp(min=lower_bound,max=upper_bound)  # range should allow decent convergence quickly
```
Comment (TheGogy) [inline src/agents/trpo/policy_network.py]:
```suggestion
        # forward_mean = max_action * tanh(1/N sum of state policies)
        for_mean = self.max_action * torch.tanh(self.mean(self.policy(state)))  

        clamped_log_std = self.log_std.clamp(min=self.lower_bound,max=self.upper_bound)  # range should allow decent convergence quickly
```

File: src/agents/trpo/record.py
Comment (TheGogy) [inline src/agents/trpo/record.py:1]: Would be good to make a version of this for TRPO specifically. Would be happy to do this tonight

File: src/agents/trpo/train.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+    # Create output dir for videos
+    video_dir = "videos/trpo"
+    os.makedirs(video_dir, exist_ok=True)
+
+    # Setup video recording
+    env = RecordVideo(
+        base_env.env,
+        video_dir,
+        episode_trigger=lambda ep: ep % 1000 == 0,
```
Comment (TheGogy) [inline src/agents/trpo/train.py]: Might be an idea to record a few more videos

```suggestion
        episode_trigger=lambda ep: ep % 100 == 0,
```

File: src/agents/trpo/train.py [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    env_info = base_env.get_env_info()
+    state_dim = env_info['observation_dim']
+    action_dim = env_info['action_dim']
+    max_action = float(env_info['action_high'][0])
+
+    agent = TRPOAgent(state_dim, action_dim, max_action, device)
+    best_reward = float('-inf')
+    running_reward = 0
+
+    agent.policy.load_state_dict(torch.load("models/trpo/trpo_best_policy.pth"))
```
Comment (TheGogy) [inline src/agents/trpo/train.py:78]: Might be an idea to leave this for the recording? The train should train an agent from scratch.

**Review by TheGogy — CHANGES_REQUESTED** [review]
LGTM for the most part! A few minor changes but should be just about good to go.

**Review by Tom-Pecher — COMMENTED** [review]
I agree with these changes. Let's get them implemented as soon as possible.

**Pull request merged**